### PR TITLE
Only set the code page for the command to run. Reset it afterwards.

### PR DESCRIPTION
**Pull request by s0undt3ch**
### What does this PR do?
See title

### CODE REVIEW

**Review by cmcmarrow — DISMISSED** [review]
APIs are better

**Comment by s0undt3ch** [comment]
And cheaper.

**Comment by markuskramerIgitt** [comment]
@s0undt3ch, great improvements.

I kindly request an additional parameter `reset_codepage` to be used like

    salt-call cmd.run “echo hello world”  reset_codepage=False


My reasoning is:

1) In my use cases, the code page of the salt-minion shell has no influence outside of that shell.

2) Decoding fails “strict” meaning that a single byte prevents the decoding of the whole communication. This is frustrating. Examples https://github.com/saltstack/salt/issues/57901 https://github.com/saltstack/salt/issues/54447

3) `stringutils.to_unicode()` uses two and only two encodings first `utf-8` and then `__salt_system_encoding__`:
- `utf-8` corresponds to code page 65001. I tested all my use cases and rolled out a (patched) minion with static code page 65001.
- `__salt_system_encoding__` can correspond to code pages 437 for the US and maybe 850 for Europe, but as you wrote above, this depends on the system administrator (which for me is an abroad third party). Adding possible other influences, differences in OS setup generations and  guess work in Python, these are many factors. Currently the use (and value) of `__salt_system_encoding__` is not revealed in the debug log and I advise to add such a debug level log entry (also in win_chcp.py) for debugging dynamic code pages. 

4) Jumping back and forth between code pages is out of my scope and does scare me. I therefore plead for an option to set the minion shell code page to utf8/65001 exactly once, like in the current code, authored by @cmcmarrow and me. 
 

To summarize: I welcome resetting the code page, but kindly request an opt-out.

Introducing an opt-out  would lead to changes e.g. in `utils/win_chcp.py`:

```
@contextmanager
def chcp(page_id, raise_error=False, reset_codepage=True):

        if set_code_page and reset_codepage:
            # Reset to the old code page
            set_page_id(previous_page_id, raise_error=raise_error)
```

Why don’t you use that function in `cmdmod._run()`?

**Comment by s0undt3ch** [comment]
> I kindly request an additional parameter reset_codepage

I think we should always reset to the previous codepage. Proper codepage handling in salt should not rely on a first command setting the codepage and not resetting it.
If our windows experts still think salt should be handling this scenario, setting the right codepage upfront, then I believe it should be done in the bat scripts used to start the salt CLI/Daemon scripts, when salt starts. Worst case, it should be set as soon as salt is imported, and not rely on `cmdmod.py` for that.

> In my use cases, the code page of the salt-minion shell has no influence outside of that shell.

What do you mean here precisely?

> Decoding fails “strict” meaning that a single byte prevents the decoding of the whole communication. This is frustrating. Examples #57901 #54447

Indeed it is and it should be fixed.

> Jumping back and forth between code pages is out of my scope and does scare me. I therefore plead for an option to set the minion shell code page to utf8/65001 exactly once, like in the current code, authored by @cmcmarrow and me.

Sure, but it should not be the `cmdmod.py` code handling that.

> Why don’t you use that function in cmdmod._run()?
It started out like that, but it would mean either:
A: the function would need to check if it's windows or not. In case it's not windows, do nothing. I don't want to account for non windows in a windows specific module. Additional code to go through when reading which might derail the reader from the function scope. Though I'm not standing ground on not doing that.
B: Have two copies of the actual subprocess call, one for windows with the context manager, one for non windows without the context manager.

This seemed like duplicating work.
As it stands now, we set it if we need to and make sure we reset it once we're done.
Also, if the desired code page is already set, we don't actually switch codepage at all.

**Comment by markuskramerIgitt** [comment]
On Sun, Sep 27, 2020 at 11:01 AM Pedro Algarvio <notifications@github.com>
wrote:

> I kindly request an additional parameter reset_codepage
>
> I think we should always reset to the previous codepage. Proper codepage
> handling in salt should not rely on a first command setting the codepage
> and not resetting it.
> If our windows experts still think salt should be handling this scenario,
> setting the right codepage upfront, then I believe it should be done in the
> bat scripts used to start the salt CLI/Daemon scripts, when salt starts.
> Worst case, it should be set as soon as salt is imported, and not rely on
> cmdmod.py for that.
>
> In my use cases, the code page of the salt-minion shell has no influence
> outside of that shell.
>
> What do you mean here precisely?
>

For years Salt sets the code page to 437 no matter what the code page was
before. In my case, the code was not 437 before.

Having Salt set the code to something different from the code page   has no
influence outside of that shell.

Only inside the shall, the code page 437 causes the two issues mentioned
below.


Decoding fails “strict” meaning that a single byte prevents the decoding of
> the whole communication. This is frustrating. Examples #57901
> <https://github.com/saltstack/salt/issues/57901> #54447
> <https://github.com/saltstack/salt/issues/54447>
>
> Indeed it is and it should be fixed.
>

Good


> Jumping back and forth between code pages is out of my scope and does
> scare me. I therefore plead for an option to set the minion shell code page
> to utf8/65001 exactly once, like in the current code, authored by
> @cmcmarrow <https://github.com/cmcmarrow> and me.
>
> Sure, but it should not be the cmdmod.py code handling that.
>

Whatever it takes to results in `salt-call cmd.run "echo nice"
--reset_windows_codepage=False  --windows_code_page=65001`

Or  `salt-call cmd.run "echo nice"  --windows_code_page=65001` because
resetting the code-page becomes an opt-in.



I am fine when I can write `salt-call cmd.run "echo nice"
--reset_codepage=False `

> Why don’t you use that function in cmdmod._run()?
> It started out like that, but it would mean either:
> A: the function would need to check if it's windows or not. In case it's
> not windows, do nothing. I don't want to account for non windows in a
> windows specific module. Additional code to go through when reading which
> might derail the reader from the function scope. Though I'm not standing
> ground on not doing that.
> B: Have two copies of the actual subprocess call, one for windows with the
> context manager, one for non windows without the context manager.
>
> This seemed like duplicating work.
> As it stands now, we set it if we need to and make sure we reset it once
> we're done.
> Also, if the desired code page is already set, we don't actually switch
> codepage at all.
>


No, as it stands now, Salt does not reset the code page.

Resetting is duplicate work, from my point of view.

I happilydedicate a Windows work group meeting to discuss. When could you
join?

**Comment by s0undt3ch** [comment]
> For years Salt sets the code page to 437 no matter what the code page was
before. In my case, the code was not 437 before.

Yes, and as you mention by the linked issues, it was not the correct approach.

> > Indeed it is and it should be fixed.

> Markus answers:
Good, where?

In this PR, but I think you're referring to the code.
Definitely, not in cmdmod.

> Whatever it takes to results in
    salt-call cmd.run "echo nice" --reset_windows_codepage=False  --windows_code_page=65001

The right solution will not involve extra CLI flags, must not, shall not. That's not the right fix.

> No, as it stands now, Salt does not reset the code page.

That's on master, and in my opinion, wrongly. If you need a codepage for a command, you should be able to set it, but it *must not* influence the commands following that one. That's unexpected behavior.

Based on the conversation I've added a commit which tries to set the codepage whenever a CLI is used, independent of what salt module is going to execute or not.
I still have to add tests to ensure that this approach does indeed solve the linked issues.

Cmdmod will still set **and** reset the codepage, **but only** if it needs to change. If the desired codepage marches the current, no switch is made.

**Review by twangboy — APPROVED** [review]
